### PR TITLE
Application Insights client

### PR DESCRIPTION
**Pull request by jeffwilcox**
This change is the early stage work to enable the user of AppInsights 
within the application, both for library functions as well as the request 
pipeline.

- Adds a new `lib/insights.js` file that adds common properties into an existing client object OR simply provides a mock-style interface with a debug provider for local test scenarios
- Moves the correlation ID assignment earlier in the middleware pipeline
- Integrates the correlation ID into event properties automatically
- Creates a wrapped/mock client with each request to pass throughout the application
- Updates the README to include the events that are being added
- Updates `lib/context.js` to use some initial ES6 goodness (`const`, `let`)
- Adds a few tests

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@jeffwilcox__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by azurecla** [comment]
Hi __@jeffwilcox__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: lib/insights.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+//
+
+'use strict';
+
+const debug = require('debug')('appinsights');
+
+function createWrappedClient(propertiesToInsert, client) {
+  const c = {};
+  if (client) {
+    c.trackEvent = addProperties.bind(client, client.trackEvent, propertiesToInsert, 1, true);
```
Comment (MichaelTsengLZ) [inline lib/insights.js]: Not quite understand this '1' here. Is it a common sense position for trackeEvent property?

Comment (jeffwilcox) [inline lib/insights.js]: Take a look at the library source code here (it is TS): https://github.com/Microsoft/ApplicationInsights-node.js/blob/master/Library/Client.ts#L49

For the `trackEvent` method, it is technically possible to have the following parameters:

- name
- properties
- measurements

Measurements are additional metrics. So in the case of that use of the function, which we have used in Witness Jobs and some other spots, the `properties` are not actually the last argument to the function, but instead the 2nd argument (index position 1).

This code is essentially [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) and injecting a little logic (a new property or properties) along the way.

**Comment by MichaelTsengLZ** [comment]
It looks perfect for me. I learnt a lot from this code. 👍

**Pull request merged**